dbms/src/WindowFunctions/IWindowFunction.cpp

```diff
@@ -31,6 +32,29 @@ namespace ErrorCodes
 extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
 } // namespace ErrorCodes
 
+namespace
+{
+bool isFrameValid(RowNumber frame_start, RowNumber frame_end, RowNumber partition_start, RowNumber partition_end)
+{
+    return !((partition_end <= frame_start) || (frame_end <= partition_start));
+}
+
+bool defaultCheckFrameValidAndHandle(WindowTransformAction & action, size_t function_index)
+{
+    if (isFrameValid(action.frame_start, action.frame_end, action.partition_start, action.partition_end))
+        return true;
+
+    IColumn & to = *action.blockAt(action.current_row).output_columns[function_index];
+    static Field null_field;
+    to.insert(null_field);
+    return false;
+}
```

Comment:
Does this only work with lead/lag? Why not as the default implementation of `checkFrameValidAndHandle`?

Reply:
> Does this only work with lead/lag? Why not as the default implementation of `checkFrameValidAndHandle`?

No, lead/lag don't need this function. So far, only first_value and last_value need this.

Because the class `WindowTransformAction` can not be used in the IWindowFunction.h, so we have to define the default function in the .cpp file.

Reply:
Is the return type of first/last must be nullable?
What if it's not nullable?

Reply:
Doesn't the frame vaild exist for other window functions?
For lead/lag, what if frame vaild?
What about other window functions in the future?

Reply:
Is there any side effect if defaultCheckFrameValidAndHandle is made the default implementation of checkFrameValidAndHandle

Reply:
> Is the return type of first/last must be nullable? What if it's not nullable?

Yes, the return type of first/last must be nullable. If it returns non-nullable, the result will be different from mysql's

Reply:
> Doesn't the frame vaild exist for other window functions? For lead/lag, what if frame vaild? What about other window functions in the future?

Some other window functions don't rely on frame such as lead/lag, so it's needless to consider frame valid for these window functions. Other window functions may have different handle method for invalid frame(maybe default handling is enough, I haven't research it yet).

Reply:
> Is there any side effect if defaultCheckFrameValidAndHandle is made the default implementation of checkFrameValidAndHandle

I think there is no side affect.